### ISSUE TITLE
ASAN build broken when using USE_ASAN=1

### ISSUE THREAD

**Issue by Flamefire** [pytorch/pytorch#41916]
## 🐛 Bug

I'm trying to debug the failures on Power by building with ASAN. Checking the documentation and combining the CONTRIBUTION.MD with .jenkins/pytorch/build-asan.sh I arrived at:
`LD_PRELOAD=${LIBASAN_RT} CC="clang" CXX="clang++" LDSHARED="clang --shared" LDFLAGS="-stdlib=libstdc++"   CFLAGS="-fsanitize=address -fsanitize=undefined -fno-sanitize-recover=all -shared-libasan -pthread"   CXX_FLAGS="-pthread -shared-libasan"   USE_ASAN=1 USE_CUDA=0 USE_MKLDNN=0 USE_OPENMP=0 BUILD_CAFFE2_OPS=0 USE_DISTRIBUTED=0 DEBUG=1  python setup.py install`

However this fails during configure with:
```
-- Performing Test HAVE_STD_REGEX
-- Performing Test HAVE_STD_REGEX
-- Performing Test HAVE_STD_REGEX -- compiled but failed to run
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
-- Performing Test HAVE_POSIX_REGEX
-- Performing Test HAVE_POSIX_REGEX
-- Performing Test HAVE_POSIX_REGEX -- compiled but failed to run
CMake Error at third_party/benchmark/CMakeLists.txt:231 (message):
  Failed to determine the source files for the regular expression backend
```

Getting some output why it failed to run I see `==109219==Your application is linked against incompatible ASan runtimes.`

I traced this back to `USE_ASAN=1` which adds flags in CMake: `"-fsanitize=address -fPIE -pie"`: https://github.com/pytorch/pytorch/blob/0ec7ba40888f07678d3a1d766817154dc90bef90/cmake/MiscCheck.cmake#L318

From the CMakeOut.log I see many lines like this:
> clang-9: warning: argument unused during compilation: '-pie' 

So it seems that clang doesn't understand `-pie` and it hence shouldn't be added unless checked before (just like other flags too)

## To Reproduce

1. Set LIBASAN_RT as described in CONTRIBUTING.md
1. Run above command

## Expected behavior

Build succeeds

## Environment

 - PyTorch Version (e.g., 1.0): 1.6.0-rc3
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Python version: 3.7.4

**Comment by mrshenli** [comment]
Adding to triage review to check if this needs to be addressed before v1.6 release.

**Comment by Flamefire** [comment]
Workaround which works locally: Instead of `CXX_FLAGS="-pthread"` use `CXXFLAGS="-fsanitize=address -fsanitize=undefined -fno-sanitize-recover=all -shared-libasan -pthread"`, the former is unused and the latter includes the required `-shared-libasan`

I further analysed this and found the following issues:
- Overwriting `CMAKE_LINKER_FLAGS_DEBUG` due to a C&P bug: https://github.com/pytorch/pytorch/blob/c0bfa45f9d26fb814a80f3053fbc550ff77d82c1/CMakeLists.txt#L596 . I recommend using `string(APPEND` instead which avoids the duplication of the variable name: https://cmake.org/cmake/help/v3.5/command/string.html#append
- Adding of `-pie` to the compile flags: https://github.com/pytorch/pytorch/blob/c0bfa45f9d26fb814a80f3053fbc550ff77d82c1/cmake/MiscCheck.cmake#L318 This is a linker flag and should be added to `CMAKE_*_LINKER_FLAGS` only, not to `CMAKE_C{XX}_FLAGS`. This may result in a false negative when checking for existing compiler flags which see the warning caused by this "unknown" compile flag
- Using of `CXX_FLAGS` environment variable instead of `CXXFLAGS`. This is likely a typo due to CMakes (arguably) better variable naming and causes it to be ignored
- Only setting (most) sanitizer flags in `CFLAGS` but not in `CXXFLAGS`
  - `-fsanitize=undefined` / UBSAN is missing
  - `-fno-sanitize-recover=all` causes sanitizer issues to be reported but not fail execution leading to a real issue: https://github.com/pytorch/pytorch/pull/41925#issuecomment-663368905
  - `-shared-libasan` is missing for C++ which leads to the bug reported here: `LD_PRELOAD` loads the shared ASAN library but for a C++ executable (such as the configure checks in GBench) a static ASAN is expected (due to `-fsanitize=address` without the `-shared-libasan`)

**Comment by Flamefire** [comment]
While testing this I found the next issue: "Your application is linked against incompatible ASan runtimes." is generated when `mkalias` is invoked. This is because sleef accidentally clears out the `CMAKE_C_FLAGS` containing the `-shared-libasan` but as `USE_ASAN=1` sets `-fsanitize=address` in the linker flags it still tries to link against ASAN  (last part is guessed)

This is due to 2 bugs in sleef which are fixed by recent PRs (2 weeks ago): https://github.com/shibatch/sleef/pull/304 and https://github.com/shibatch/sleef/pull/312

I'd recommend updating the submodule

**Comment by gchanan** [comment]
I don't think this needs to be in 1.6.0 -- the first is getting a proper setup, the second are sleef issues that just got fixed (so can wait a release).
